### PR TITLE
Add validation for new file names for sql projects

### LINKED ISSUE #17654: Validation for new file names for sql projects
There currently isn't any validation for the new file names of items added to sql projects. For reference, we previously added file name validation in the dacpac extension https://github.com/microsoft/azuredatastudio/blob/main/extensions/dacpac/src/wizard/api/utils.ts#L53

We should have validation similar to in VS:
image

### PR DESCRIPTION
**Pull request by SakshiS-harma**
This PR fixes #17654 and adds validation for new file names in sql projects.

Validation for new project created from database.
projectValidation

Validation for new items added to the project
newItem

Validation for new project
newProject

### CODE REVIEW

File: extensions/sql-database-projects/src/test/dialogs/utils.test.ts [hunk 1]
```diff
@@ -30,3 +34,164 @@ describe('Tests to verify dialog utils functions', function (): void {
 	});
 });
 
+describe('Sanitize database name for filename tests', function (): void {
```
Comment (kisantia) [inline extensions/sql-database-projects/src/test/dialogs/utils.test.ts:37]: nice test coverage with all the different cases!

**Review by kisantia — COMMENTED** [review]
please add screenshots or gifs of where this new validation was added and how it looks (ex: dialogs, quickpicks)

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 2]
```diff
@@ -660,7 +661,10 @@ export class ProjectsController {
 
 		const itemObjectName = await vscode.window.showInputBox({
 			prompt: constants.newObjectNamePrompt(itemType.friendlyName),
-			value: `${suggestedName}${counter}`,
+			value: sanitizeStringForFilename(`${suggestedName}${counter}`),
```
[the remark this answers is not in the document]
Comment (SakshiS-harma) [reply extensions/sql-database-projects/src/controllers/projectController.ts]: Done.

**Comment by SakshiS-harma** [comment]
> please add screenshots or gifs of where this new validation was added and how it looks (ex: dialogs, quickpicks)

Added gifs for the validation. Still need to add quickpick gifs for sql-project-vscode.

File: extensions/data-workspace/src/dialogs/utils.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+const UNIX_INVALID_FILE_CHARS = /[\\/]/g;
+const isWindows = os.platform() === 'win32';
+const WINDOWS_FORBIDDEN_NAMES = /^(con|prn|aux|clock\$|nul|lpt[0-9]|com[0-9])$/i;
+
+/**
+ * Returns true if the string is a valid filename
+ * Logic is copied from src\vs\base\common\extpath.ts
+ * @param name filename to check
+ */
+export function isValidBasename(name: string | null | undefined): boolean {
```
Comment (kisantia) [inline extensions/data-workspace/src/dialogs/utils.ts:22]: ideally these validation functions don't need to be added in each of these extensions and could be reused from one common place. Can this just get added in the data workspace extension, then expose it in data-workspace.d.ts so the sql projects extension can use it?

Comment (Benjin) [reply extensions/data-workspace/src/dialogs/utils.ts:22]: If this doesn't get moved to dataworkspace and exposed, then I think src/common/pathUtils.ts would be a more appropriate place than within the dialogs folder.

**Review by Benjin — CHANGES_REQUESTED** [review]

**Comment by Benjin** [comment]
VS only ran on Windows, so there was no concern of cross-platform portability of project/object names.  Now that we're operating cross-platform, do we want to give block or give warnings when choosing a name that works on your current platform but wouldn't for the other?

File: extensions/data-workspace/src/dialogs/utils.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
+/*---------------------------------------------------------------------------------------------
```
Comment (Benjin) [inline extensions/data-workspace/src/dialogs/utils.ts:1]: VS only ran on Windows so there was no concern of cross-platform compatibility.  Now that we run on Unix and Windows, do we want to block or warn when choosing a name that wouldn't work on the other platform?

Comment (SakshiS-harma) [reply extensions/data-workspace/src/dialogs/utils.ts:1]: In the changes, it does take into consideration for invalid file chars in Windows and Unix (line 12). So it does character validation based on the platform.

File: extensions/data-workspace/src/dialogs/utils.ts (hunk 3 above)
Comment (SakshiS-harma) [reply extensions/data-workspace/src/dialogs/utils.ts]: Removed the code in projects extension. And moved the code under src/comon/pathUtils.

File: extensions/data-workspace/src/test/dialogs/utils.test.ts [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+		should(isValidBasenameErrorMessage(formatFileName('".sqlproj'))).equal(isWindows ? constants.invalidFileCharsErrorMessage : '');
+
+		// Windows filenames cannot end with a whitespace
+		should(isValidBasenameErrorMessage(formatFileName('test   .sqlproj'))).equal(isWindows ? constants.trailingWhitespaceErrorMessage : '');
+		should(isValidBasenameErrorMessage(formatFileName('test	.sqlproj'))).equal(isWindows ? constants.trailingWhitespaceErrorMessage : '');
+	});
+
+	it('Should determine Windows forbidden filenames', async () => {
+		// invalid only for Windows
+		should(isValidBasenameErrorMessage(formatFileName('CON.sqlproj'))).equal(isWindows ? constants.reservedWindowsFilenameErrorMessage : '');
```
[the remark this answers is not in the document]
Comment (SakshiS-harma) [reply extensions/data-workspace/src/test/dialogs/utils.test.ts]: Updated tests.

File: extensions/data-workspace/src/common/dataWorkspaceExtension.ts [hunk 6]
```diff
@@ -41,4 +42,17 @@ export class DataWorkspaceExtension implements IExtension {
 		return openSpecificProjectNewProjectDialog(projectType, this.workspaceService);
 	}
 
+	isValidFilenameCharacter(c: string): boolean {
+		return isValidFilenameCharacter(c);
+	}
+	sanitizeStringForFilename(s: string): string {
+		return sanitizeStringForFilename(s);
+	}
+	isValidBasename(name: string | null | undefined): boolean {
```
Comment (kisantia) [inline extensions/data-workspace/src/common/dataWorkspaceExtension.ts:51]: this parameter should be able to be simplified to `name?: string` and below

File: extensions/data-workspace/src/common/dataWorkspaceExtension.ts [hunk 7]
```diff
@@ -41,4 +42,17 @@ export class DataWorkspaceExtension implements IExtension {
 		return openSpecificProjectNewProjectDialog(projectType, this.workspaceService);
 	}
 
+	isValidFilenameCharacter(c: string): boolean {
+		return isValidFilenameCharacter(c);
+	}
+	sanitizeStringForFilename(s: string): string {
```
Comment (kisantia) [inline extensions/data-workspace/src/common/dataWorkspaceExtension.ts:48]: nit: blank lines between these functions

File: extensions/sql-database-projects/src/test/dialogs/utils.test.ts [hunk 8]
```diff
@@ -1,32 +0,0 @@
...
- *  Copyright (c) Microsoft Corporation. All rights reserved.
- *  Licensed under the Source EULA. See License.txt in the project root for license information.
- *--------------------------------------------------------------------------------------------*/
-
-import * as should from 'should';
-import * as constants from '../../common/constants';
-import { SqlTargetPlatform } from 'sqldbproj';
-import { getDefaultDockerImageWithTag, getDockerBaseImages } from '../../dialogs/utils';
-
-describe('Tests to verify dialog utils functions', function (): void {
```
Comment (kisantia) [inline extensions/sql-database-projects/src/test/dialogs/utils.test.ts:11]: this file shouldn't be deleted, right?

Comment (SakshiS-harma) [reply extensions/sql-database-projects/src/test/dialogs/utils.test.ts:11]: I have these tests added in data-workspace extension, hence removed this file from here.

Comment (kisantia) [reply extensions/sql-database-projects/src/test/dialogs/utils.test.ts:11]: There were some docker tests in this file that shouldn't be deleted

Comment (SakshiS-harma) [reply extensions/sql-database-projects/src/test/dialogs/utils.test.ts:11]: Sorry, my bad. I forgot there were tests up as well, I will place it back.

**Review by kisantia — APPROVED** [review]

**Pull request merged**